File: ports/1.21.1-full/src/main/java/mods/eln/node/simple/SimpleNodeItem.java
package mods.eln.node.simple;

import mods.eln.misc.Coordinate;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.core.Direction;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;

public class SimpleNodeItem extends BlockItem {
    SimpleNodeBlock block;

    public SimpleNodeItem(Block b) {
        super(b);
        block = (SimpleNodeBlock) b;
    }

    @Override
    public boolean placeBlockAt(ItemStack stack, Player player, Level world, BlockPos pos, Direction side, float hitX, float hitY, float hitZ, BlockState newState) {
        SimpleNode node = null;
        if (!world.isRemote) {
            node = block.newNode();
            node.setDescriptorKey(block.descriptorKey);
            node.onBlockPlacedBy(new Coordinate(pos, world), block.getFrontForPlacement(player), player, stack);
        }

        if (!world.setBlockState(pos, newState, 3)) {
            if (node != null) node.onBreakBlock();
            return false;
        }

        BlockState state = world.getBlockState(pos);
        if (state.getBlock() == this.block) {
            this.block.onBlockPlacedBy(world, pos, state, player, stack);
        }

        return true;
    }
}

File: src/main/kotlin/mods/eln/node/simple/SimpleNodeItem.kt
package mods.eln.node.simple

import mods.eln.misc.Coordinate
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.context.BlockPlaceContext
import mods.eln.misc.getBlockState

class SimpleNodeItem(b: Block, properties: Properties = Properties()) : BlockItem(b, properties) {
    val nodeBlock: SimpleNodeBlock = b as SimpleNodeBlock

    /**
     * The node has to exist before the block is placed: SimpleNodeEntity looks its node up by
     * coordinate as soon as it is created, and a placement that fails must not leave one behind.
     * (1.7.10's `placeBlockAt`; vanilla's `place` does the sound, the stat and the stack around it.)
     */
    override fun placeBlock(context: BlockPlaceContext, newState: BlockState): Boolean {
        val world = context.level
        val pos = context.clickedPos
        val player = context.player ?: return false
        var node: SimpleNode? = null
        if (!world.isClientSide) {
            node = nodeBlock.newNode()
            node!!.descriptorKey = nodeBlock.descriptorKey
            node.onBlockPlacedBy(Coordinate(pos.x, pos.y, pos.z, world), nodeBlock.getFrontForPlacement(player), player, context.itemInHand)
        }
        if (!super.placeBlock(context, newState)) {
            node?.onBreakBlock()
            return false
        }
        return true
    }
}
